### PR TITLE
Refactor and fix container streaming.

### PR DESCRIPTION
**Pull request by Random-Liu**
This PR:
1) Cleanup and refactor container streaming code.
2) Add our own IO pipe logic for exec. The new exec io pipe logic has similar behavior with attach. (Actually in docker, exec and attach uses the same piece of io pipe code, https://github.com/moby/moby/blob/master/container/stream/attach.go#L60).

This fixes the e2e test timeout. https://k8s-testgrid.appspot.com/sig-node-containerd#e2e-gci
The test `Kubectl should handle in-cluster config` timed out, because it assumes that after input ends with EOF, exec stdin should be closed, and `tar xvf - XXX` should stop. 
However, in default containerd io pipe, even if stdio finishes, it still holds the container stdio open. So the test blocks on the following line forever:
```
INFO: Running '/workspace/github.com/kubernetes-incubator/cri-containerd/kubernetes/platforms/linux/amd64/kubectl --server=https://104.197.32.31 --kubeconfig=/workspace/.kube/config cp /workspace/github.com/kubernetes-incubator/cri-containerd/kubernetes/platforms/linux/amd64/kubectl e2e-tests-kubectl-hzzv7/nginx:/tmp/'
```
With this PR, we'll close exec stdin if input finishes. It should fix the test timeout.

Signed-off-by: Lantao Liu <lantaol@google.com>

### CODE REVIEW

File: pkg/server/io/container_io.go [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+	return func(c *ContainerIO) error {
+		c.fifos = fifos
+		return nil
+	}
+}
+
+// WithNewFIFOs creates new fifos for the container io.
+func WithNewFIFOs(root string, tty, stdin bool) ContainerIOOpts {
+	return func(c *ContainerIO) error {
+		fifos, err := newFifos(root, c.id, tty)
```
[the remark this answers is not in the document]
Comment (Random-Liu) [reply pkg/server/io/container_io.go]: Good suggestion!

File: pkg/server/container_execsync.go [hunk 2]
```diff
@@ -163,13 +176,17 @@ func (c *criContainerdService) execInContainer(ctx context.Context, id string, o
 		exitRes := <-exitCh
 		glog.V(2).Infof("Timeout received while waiting for exec process kill %q code %d and error %v",
 			execID, exitRes.ExitCode(), exitRes.Error())
+		<-attachDone
```
Comment (mikebrow) [inline pkg/server/container_execsync.go]: // wait for the attachDone channel to be closed

File: pkg/server/io/exec_io.go [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+	return containerd.IOConfig{
+		Terminal: e.fifos.Terminal,
+		Stdin:    e.fifos.In,
+		Stdout:   e.fifos.Out,
+		Stderr:   e.fifos.Err,
+	}
+}
+
+// Attach attaches exec stdio. The logic is similar with container io attach.
+func (e *ExecIO) Attach(opts AttachOptions) <-chan struct{} {
```
Comment (mikebrow) [inline pkg/server/io/exec_io.go]: // returns a `done` channel that will be closed when all streams have closed

File: pkg/server/io/exec_io.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+		wg.Add(1)
+		// Closer should wait for this routine to be over.
+		e.closer.wg.Add(1)
+		go attachOutput(Stderr, opts.Stderr, e.stderr)
+	}
+
+	done := make(chan struct{})
+	go func() {
+		wg.Wait()
+		close(done)
```
Comment (mikebrow) [inline pkg/server/io/exec_io.go]: // wake up the threads that are waiting for the done channel to be closed.

File: pkg/server/io/container_io.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (Random-Liu) [reply pkg/server/io/container_io.go]: Done.

File: pkg/server/io/container_io.go [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+		if err != nil {
+			return err
+		}
+		c.fifos = fifos
```
[the remark this answers is not in the document]
Comment (Random-Liu) [reply pkg/server/io/container_io.go]: Done

File: pkg/server/io/container_io.go [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+		if !stdin {
```
[the remark this answers is not in the document]
Comment (Random-Liu) [reply pkg/server/io/container_io.go]: Done

**Comment by Random-Liu** [comment]
@mikebrow Addressed comments. Thanks for reviewing! Very good cleanup suggestion. :)

**Pull request merged**